File: app/src/main/java/ceui/lisa/activities/MainActivity.java
package ceui.lisa.activities;

import android.Manifest;
import android.annotation.SuppressLint;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.Color;
import android.os.Bundle;
import android.provider.MediaStore;
import android.text.TextUtils;
import android.view.KeyEvent;
import android.view.MenuItem;
import android.view.MotionEvent;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.core.view.GravityCompat;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentPagerAdapter;
import androidx.viewpager.widget.ViewPager;

import com.bumptech.glide.Glide;
import com.google.android.material.bottomnavigation.BottomNavigationView;
import com.google.android.material.navigation.NavigationView;
import com.qmuiteam.qmui.widget.dialog.QMUIDialog;

import ceui.lisa.R;
import ceui.lisa.core.Manager;
import ceui.lisa.databinding.ActivityCoverBinding;
import ceui.lisa.fragments.FragmentCenter;
import ceui.lisa.fragments.FragmentLeft;
import ceui.lisa.fragments.FragmentRight;
import ceui.lisa.fragments.FragmentViewPager;
import ceui.lisa.helper.DrawerLayoutHelper;
import ceui.lisa.helper.NavigationLocationHelper;
import ceui.lisa.utils.Common;
import ceui.lisa.utils.GlideUtil;
import ceui.lisa.utils.Params;
import ceui.lisa.view.DrawerLayoutViewPager;

import static ceui.lisa.R.id.nav_gallery;
import static ceui.lisa.R.id.nav_slideshow;
import static ceui.lisa.activities.Shaft.sUserModel;

/**
 * 主页
 */
public class MainActivity extends BaseActivity<ActivityCoverBinding>
        implements NavigationView.OnNavigationItemSelectedListener {

    private static final int REQUEST_WRITE_STORAGE = 1001;
    private static final String EXTRA_REFRESH_DRAWER_HEADER = "refreshDrawerHeader";
    private static final int PAGE_LEFT = 0;
    private static final int PAGE_CENTER = 1;
    private static final int PAGE_RIGHT = 2;
    private static final int PAGE_R18 = 3;

    private ImageView userHead;
    private TextView username;
    private TextView user_email;
    private long mExitTime;
    private Fragment[] baseFragments = null;

    @Override
    protected int initLayout() {
        return R.layout.activity_cover;
    }

    @Override
    public boolean hideStatusBar() {
        return true;
    }

    @Override
    protected void initView() {
        getBaseBind().drawerLayout.setScrimColor(Color.TRANSPARENT);
        getBaseBind().navView.setNavigationItemSelectedListener(this);
        userHead = getBaseBind().navView.getHeaderView(0).findViewById(R.id.user_head);
        username = getBaseBind().navView.getHeaderView(0).findViewById(R.id.user_name);
        user_email = getBaseBind().navView.getHeaderView(0).findViewById(R.id.user_email);
        initDrawerHeader();
        userHead.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Common.showUser(getMContext(), sUserModel);
                getBaseBind().drawerLayout.closeDrawer(GravityCompat.START);
            }
        });
        userHead.setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View v) {
                boolean filterEnable = Shaft.sSettings.isR18FilterTempEnable();
                Shaft.sSettings.setR18FilterTempEnable(!filterEnable);
                Common.showToast(filterEnable ? "ԅ(♡﹃♡ԅ)" : "X﹏X");
                return true;
            }
        });
        getBaseBind().navigationView.setOnNavigationItemSelectedListener(new BottomNavigationView.OnNavigationItemSelectedListener() {
            @Override
            public boolean onNavigationItemSelected(@NonNull MenuItem item) {
                return navigateToPage(item.getItemId());
            }
        });
        getBaseBind().navigationView.setOnNavigationItemReselectedListener(new BottomNavigationView.OnNavigationItemReselectedListener() {
            @Override
            public void onNavigationItemReselected(@NonNull MenuItem item) {
                forceRefreshCurrentPage(item.getItemId());
            }
        });
        getBaseBind().viewPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {

            }

            @Override
            public void onPageSelected(int position) {
                syncBottomNavigation(position);
            }

            @Override
            public void onPageScrollStateChanged(int state) {

            }
        });

        getBaseBind().viewPager.setTouchEventForwarder(new DrawerLayoutViewPager.IForwardTouchEvent() {
            @Override
            public void forwardTouchEvent(MotionEvent ev) {
                getDrawer().onTouchEvent(ev);
            }
        });
        DrawerLayoutHelper.setCustomLeftEdgeSize(getDrawer(), 1.0f);
    }

    public static Intent newIntent(Context context, boolean refreshDrawerHeader) {
        Intent intent = new Intent(context, MainActivity.class);
        intent.putExtra(EXTRA_REFRESH_DRAWER_HEADER, refreshDrawerHeader);
        return intent;
    }

    private void initFragment() {
        if (Shaft.sSettings.isMainViewR18()) {
            getBaseBind().navigationView.inflateMenu(R.menu.main_activity0_with_r18);
            baseFragments = new Fragment[]{
                    new FragmentLeft(),
                    new FragmentCenter(),
                    new FragmentRight(),
                    FragmentViewPager.newInstance(Params.VIEW_PAGER_R18),
            };
        } else {
            getBaseBind().navigationView.inflateMenu(R.menu.main_activity0);
            baseFragments = new Fragment[]{
                    new FragmentLeft(),
                    new FragmentCenter(),
                    new FragmentRight()
            };
        }
        getBaseBind().viewPager.setAdapter(new FragmentPagerAdapter(getSupportFragmentManager()) {
            @Override
            public Fragment getItem(int i) {
                return baseFragments[i];
            }

            @Override
            public int getCount() {
                return baseFragments.length;
            }
        });
        getBaseBind().viewPager.setOffscreenPageLimit(baseFragments.length - 1);
        getBaseBind().viewPager.setCurrentItem(getNavigationInitPosition());
        Manager.get().restoreAsync();
    }

    @Override
    protected void initData() {
        if (!hasLoggedInUser()) {
            redirectToLogin();
            return;
        }
        continueStartup();
    }

    private boolean hasLoggedInUser() {
        return sUserModel != null
                && sUserModel.getUser() != null
                && sUserModel.getUser().isIs_login();
    }

    private void redirectToLogin() {
        TemplateActivity.startLogin(getMContext());
        finish();
    }

    private void continueStartup() {
        if (shouldSkipStoragePermission()) {
            completeStartup();
            return;
        }
        ensureStoragePermission();
    }

    private boolean shouldSkipStoragePermission() {
        return Common.isAndroidQ();
    }

    private void completeStartup() {
        initFragment();
    }

    private void ensureStoragePermission() {
        if (hasStoragePermission()) {
            completeStartup();
            return;
        }
        requestStoragePermission();
    }

    private boolean hasStoragePermission() {
        return ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE)
                == android.content.pm.PackageManager.PERMISSION_GRANTED;
    }

    private void requestStoragePermission() {
        ActivityCompat.requestPermissions(
                this,
                new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE},
                REQUEST_WRITE_STORAGE
        );
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == REQUEST_WRITE_STORAGE) {
            handleStoragePermissionResult(grantResults);
        }
    }

    private void handleStoragePermissionResult(@NonNull int[] grantResults) {
        if (isPermissionGranted(grantResults)) {
            completeStartup();
            return;
        }
        handleStoragePermissionDenied();
    }

    private boolean isPermissionGranted(@NonNull int[] grantResults) {
        return grantResults.length > 0
                && grantResults[0] == android.content.pm.PackageManager.PERMISSION_GRANTED;
    }

    private void handleStoragePermissionDenied() {
        Common.showToast(getString(R.string.access_denied));
        finish();
    }

    public DrawerLayout getDrawer() {
        return getBaseBind().drawerLayout;
    }

    @SuppressLint("NonConstantResourceId")
    @Override
    public boolean onNavigationItemSelected(MenuItem item) {
        handleDrawerNavigation(item.getItemId());
        getBaseBind().drawerLayout.closeDrawer(GravityCompat.START);
        return true;
    }

    private void handleDrawerNavigation(int itemId) {
        if (itemId == R.id.nav_new_work) {
            TemplateActivity.startLatest(this);
            return;
        }
        Intent intent = resolveDrawerDestination(itemId);
        if (intent != null) {
            startActivity(intent);
        }
    }

    private boolean navigateToPage(int itemId) {
        int pageIndex = getPageIndex(itemId);
        if (pageIndex == -1) {
            return false;
        }
        getBaseBind().viewPager.setCurrentItem(pageIndex);
        return true;
    }

    private int getPageIndex(int itemId) {
        if (itemId == R.id.action_1) {
            return PAGE_LEFT;
        } else if (itemId == R.id.action_2) {
            return PAGE_CENTER;
        } else if (itemId == R.id.action_3) {
            return PAGE_RIGHT;
        } else if (itemId == R.id.action_4) {
            return PAGE_R18;
        }
        return -1;
    }

    private void forceRefreshCurrentPage(int itemId) {
        int pageIndex = getPageIndex(itemId);
        if (pageIndex == -1) {
            return;
        }
        forceRefreshPage(pageIndex);
    }

    private void forceRefreshPage(int pageIndex) {
        Fragment fragment = getBaseFragment(pageIndex);
        if (fragment instanceof FragmentLeft) {
            ((FragmentLeft) fragment).forceRefresh();
        } else if (fragment instanceof FragmentCenter) {
            ((FragmentCenter) fragment).forceRefresh();
        } else if (fragment instanceof FragmentRight) {
            ((FragmentRight) fragment).forceRefresh();
        } else if (fragment instanceof FragmentViewPager) {
            ((FragmentViewPager) fragment).forceRefresh();
        }
    }

    private Fragment getBaseFragment(int pageIndex) {
        if (pageIndex < 0 || pageIndex >= baseFragments.length) {
            return null;
        }
        return baseFragments[pageIndex];
    }

    private void syncBottomNavigation(int position) {
        int itemId = getBottomNavigationItemId(position);
        if (itemId != View.NO_ID) {
            getBaseBind().navigationView.setSelectedItemId(itemId);
        }
    }

    private int getBottomNavigationItemId(int pageIndex) {
        if (pageIndex == PAGE_LEFT) {
            return R.id.action_1;
        } else if (pageIndex == PAGE_CENTER) {
            return R.id.action_2;
        } else if (pageIndex == PAGE_RIGHT) {
            return R.id.action_3;
        } else if (pageIndex == PAGE_R18) {
            return R.id.action_4;
        }
        return View.NO_ID;
    }

    private Intent resolveDrawerDestination(int itemId) {
        if (itemId == nav_gallery) {
            return TemplateActivity.newDownloadManagerIntent(this);
        } else if (itemId == nav_slideshow) {
            return TemplateActivity.newHistoryIntent(this);
        } else if (itemId == R.id.nav_manage) {
            return TemplateActivity.newSettingsIntent(this);
        } else if (itemId == R.id.nav_share) {
            return TemplateActivity.newAboutIntent(this);
        } else if (itemId == R.id.main_page) {
            return UActivity.newCurrentUserIntent(this);
        } else if (itemId == R.id.muted_list) {
            return TemplateActivity.newMutedTagsIntent(this);
        } else if (itemId == R.id.nav_feature) {
            return TemplateActivity.newFeatureIntent(this);
        } else if (itemId == R.id.nav_fans) {
            return TemplateActivity.newFansIntent(this, sUserModel.getUser().getId());
        } else if (itemId == R.id.illust_star) {
            return TemplateActivity.newMyIllustBookmarksIntent(this);
        } else if (itemId == R.id.novel_star) {
            return TemplateActivity.newMyNovelBookmarksIntent(this);
        } else if (itemId == R.id.watchlist) {
            return TemplateActivity.newWatchlistIntent(this);
        } else if (itemId == R.id.novel_markers) {
            return TemplateActivity.newNovelMarkersIntent(this);
        } else if (itemId == R.id.follow_user) {
            return TemplateActivity.newMyFollowingIntent(this);
        }
        return null;
    }

    @Override
    protected void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.clear();
    }

    private void initDrawerHeader() {
        if (sUserModel != null && sUserModel.getUser() != null) {
            Glide.with(getMContext())
                    .load(GlideUtil.getHead(sUserModel.getUser()))
                    .into(userHead);
            username.setText(sUserModel.getUser().getName());
            user_email.setText(TextUtils.isEmpty(sUserModel.getUser().getMail_address()) ?
                    getMContext().getString(R.string.no_mail_address) : sUserModel.getUser().getMail_address());
        }
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if (handleOpenDrawerOnBack()) {
            return true;
        }
        if (shouldInterceptBack(keyCode, event)) {
            exit();
            return true;
        }
        return false;
    }

    private boolean handleOpenDrawerOnBack() {
        if (!getBaseBind().drawerLayout.isDrawerOpen(GravityCompat.START)) {
            return false;
        }
        getBaseBind().drawerLayout.closeDrawer(GravityCompat.START);
        return true;
    }

    private boolean shouldInterceptBack(int keyCode, KeyEvent event) {
        return keyCode == KeyEvent.KEYCODE_BACK && event.getRepeatCount() == 0;
    }

    public void exit() {
        if (shouldFinishOnBackPress()) {
            finish();
            return;
        }
        if (hasPendingDownloadTasks()) {
            showDownloadTaskExitDialog();
            return;
        }
        recordPendingExit();
    }

    private boolean shouldFinishOnBackPress() {
        return (System.currentTimeMillis() - mExitTime) <= 2000;
    }

    private boolean hasPendingDownloadTasks() {
        return Manager.get().getContent().size() != 0;
    }

    private void showDownloadTaskExitDialog() {
        AlertDialog.Builder builder = new AlertDialog.Builder(getMContext());
        builder.setTitle(getString(R.string.shaft_hint));
        builder.setMessage(getMContext().getString(R.string.you_have_download_plan));
        builder.setPositiveButton(getMContext().getString(R.string.sure), new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                stopDownloadsAndExit();
            }
        });
        builder.setNegativeButton(getMContext().getString(R.string.cancel), null);
        builder.setNeutralButton(getString(R.string.see_download_task), (dialog, which) -> openDownloadManager());
        builder.create().show();
    }

    private void stopDownloadsAndExit() {
        Manager.get().stopAll();
        finish();
    }

    private void openDownloadManager() {
        TemplateActivity.startDownloadManager(this);
    }

    private void recordPendingExit() {
        Common.showToast(getString(R.string.double_click_finish));
        mExitTime = System.currentTimeMillis();
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (shouldRefreshDrawerHeader()) {
            initDrawerHeader();
            getIntent().removeExtra(EXTRA_REFRESH_DRAWER_HEADER);
        }
    }

    private boolean shouldRefreshDrawerHeader() {
        return getIntent().getBooleanExtra(EXTRA_REFRESH_DRAWER_HEADER, false);
    }

    @Override
    public void finish() {
        int currentPosition = getBaseBind().viewPager.getCurrentItem();
        Shaft.getMMKV().putInt(Params.MAIN_ACTIVITY_NAVIGATION_POSITION, currentPosition);
        super.finish();
    }

    private int getNavigationInitPosition() {
        int defaultPosition = 0;
        String settingValue = Shaft.sSettings.getNavigationInitPosition();
        if (settingValue.equals(NavigationLocationHelper.LATEST)) {
            int latestPosition = Shaft.getMMKV().getInt(Params.MAIN_ACTIVITY_NAVIGATION_POSITION, 0);
            return latestPosition < baseFragments.length ? latestPosition : defaultPosition;
        }
        NavigationLocationHelper.NavigationItem navigationValue = NavigationLocationHelper.NAVIGATION_MAP.getOrDefault(settingValue, null);
        if (navigationValue == null) {
            return defaultPosition;
        }
        Class clazz = navigationValue.getInstanceClass();
        for (int i = 0; i < baseFragments.length; i++) {
            Fragment fragment = baseFragments[i];
            if (clazz == fragment.getClass()) {
                return i;
            }
        }
        return defaultPosition;
    }
}

File: app/src/main/java/ceui/lisa/activities/MainActivity.kt
package ceui.lisa.activities

import android.Manifest
import android.annotation.SuppressLint
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Color
import android.os.Bundle
import android.text.TextUtils
import android.view.KeyEvent
import android.view.MenuItem
import android.view.MotionEvent
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.core.view.GravityCompat
import androidx.drawerlayout.widget.DrawerLayout
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentPagerAdapter
import androidx.viewpager.widget.ViewPager
import ceui.lisa.R
import ceui.lisa.core.Manager
import ceui.lisa.databinding.ActivityCoverBinding
import ceui.lisa.fragments.FragmentCenter
import ceui.lisa.fragments.FragmentLeft
import ceui.lisa.fragments.FragmentRight
import ceui.lisa.fragments.FragmentViewPager
import ceui.lisa.helper.DrawerLayoutHelper
import ceui.lisa.helper.NavigationLocationHelper
import ceui.lisa.utils.Common
import ceui.lisa.utils.GlideUtil
import ceui.lisa.utils.Params
import ceui.lisa.view.DrawerLayoutViewPager
import com.bumptech.glide.Glide
import com.google.android.material.bottomnavigation.BottomNavigationView
import com.google.android.material.navigation.NavigationView

/**
 * 主页
 */
class MainActivity : BaseActivity<ActivityCoverBinding>(), NavigationView.OnNavigationItemSelectedListener {

    private var userHead: ImageView? = null
    private var username: TextView? = null
    private var userEmail: TextView? = null
    private var mExitTime = 0L
    private lateinit var baseFragments: Array<Fragment>

    override fun initLayout(): Int = R.layout.activity_cover

    override fun hideStatusBar(): Boolean = true

    override fun initView() {
        baseBind.drawerLayout.setScrimColor(Color.TRANSPARENT)
        baseBind.navView.setNavigationItemSelectedListener(this)
        userHead = baseBind.navView.getHeaderView(0).findViewById(R.id.user_head)
        username = baseBind.navView.getHeaderView(0).findViewById(R.id.user_name)
        userEmail = baseBind.navView.getHeaderView(0).findViewById(R.id.user_email)
        initDrawerHeader()
        userHead?.setOnClickListener {
            Common.showUser(mContext, Shaft.sUserModel)
            baseBind.drawerLayout.closeDrawer(GravityCompat.START)
        }
        userHead?.setOnLongClickListener {
            val filterEnable = Shaft.sSettings.isR18FilterTempEnable()
            Shaft.sSettings.setR18FilterTempEnable(!filterEnable)
            Common.showToast(if (filterEnable) "ԅ(♡﹃♡ԅ)" else "X﹏X")
            true
        }
        baseBind.navigationView.setOnNavigationItemSelectedListener { item ->
            navigateToPage(item.itemId)
        }
        baseBind.navigationView.setOnNavigationItemReselectedListener { item ->
            forceRefreshCurrentPage(item.itemId)
        }
        baseBind.viewPager.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
            override fun onPageScrolled(
                position: Int,
                positionOffset: Float,
                positionOffsetPixels: Int,
            ) {
            }

            override fun onPageSelected(position: Int) {
                syncBottomNavigation(position)
            }

            override fun onPageScrollStateChanged(state: Int) {
            }
        })
        baseBind.viewPager.setTouchEventForwarder(object : DrawerLayoutViewPager.IForwardTouchEvent {
            override fun forwardTouchEvent(ev: MotionEvent) {
                getDrawer().onTouchEvent(ev)
            }
        })
        DrawerLayoutHelper.setCustomLeftEdgeSize(getDrawer(), 1.0f)
    }

    private fun initFragment() {
        if (Shaft.sSettings.isMainViewR18()) {
            baseBind.navigationView.inflateMenu(R.menu.main_activity0_with_r18)
            baseFragments = arrayOf(
                FragmentLeft(),
                FragmentCenter(),
                FragmentRight(),
                FragmentViewPager.newInstance(Params.VIEW_PAGER_R18),
            )
        } else {
            baseBind.navigationView.inflateMenu(R.menu.main_activity0)
            baseFragments = arrayOf(
                FragmentLeft(),
                FragmentCenter(),
                FragmentRight(),
            )
        }
        baseBind.viewPager.adapter = object : FragmentPagerAdapter(supportFragmentManager) {
            override fun getItem(position: Int): Fragment = baseFragments[position]

            override fun getCount(): Int = baseFragments.size
        }
        baseBind.viewPager.offscreenPageLimit = baseFragments.size - 1
        baseBind.viewPager.currentItem = getNavigationInitPosition()
        Manager.get().restoreAsync()
    }

    override fun initData() {
        if (!hasLoggedInUser()) {
            redirectToLogin()
            return
        }
        continueStartup()
    }

    private fun hasLoggedInUser(): Boolean {
        val user = Shaft.sUserModel.getUser()
        return user != null && user.isIs_login()
    }

    private fun redirectToLogin() {
        TemplateActivity.startLogin(mContext)
        finish()
    }

    private fun continueStartup() {
        if (shouldSkipStoragePermission()) {
            completeStartup()
            return
        }
        ensureStoragePermission()
    }

    private fun shouldSkipStoragePermission(): Boolean = Common.isAndroidQ()

    private fun completeStartup() {
        initFragment()
    }

    private fun ensureStoragePermission() {
        if (hasStoragePermission()) {
            completeStartup()
            return
        }
        requestStoragePermission()
    }

    private fun hasStoragePermission(): Boolean {
        return ContextCompat.checkSelfPermission(
            this,
            Manifest.permission.WRITE_EXTERNAL_STORAGE,
        ) == PackageManager.PERMISSION_GRANTED
    }

    private fun requestStoragePermission() {
        ActivityCompat.requestPermissions(
            this,
            arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE),
            REQUEST_WRITE_STORAGE,
        )
    }

    @Deprecated("Deprecated in Java")
    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray,
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == REQUEST_WRITE_STORAGE) {
            handleStoragePermissionResult(grantResults)
        }
    }

    private fun handleStoragePermissionResult(grantResults: IntArray) {
        if (isPermissionGranted(grantResults)) {
            completeStartup()
            return
        }
        handleStoragePermissionDenied()
    }

    private fun isPermissionGranted(grantResults: IntArray): Boolean {
        return grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED
    }

    private fun handleStoragePermissionDenied() {
        Common.showToast(getString(R.string.access_denied))
        finish()
    }

    fun getDrawer(): DrawerLayout = baseBind.drawerLayout

    @SuppressLint("NonConstantResourceId")
    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        handleDrawerNavigation(item.itemId)
        baseBind.drawerLayout.closeDrawer(GravityCompat.START)
        return true
    }

    private fun handleDrawerNavigation(itemId: Int) {
        if (itemId == R.id.nav_new_work) {
            TemplateActivity.startLatest(this)
            return
        }
        val intent = resolveDrawerDestination(itemId)
        if (intent != null) {
            startActivity(intent)
        }
    }

    private fun navigateToPage(itemId: Int): Boolean {
        val pageIndex = getPageIndex(itemId)
        if (pageIndex == -1) {
            return false
        }
        baseBind.viewPager.currentItem = pageIndex
        return true
    }

    private fun getPageIndex(itemId: Int): Int {
        return when (itemId) {
            R.id.action_1 -> PAGE_LEFT
            R.id.action_2 -> PAGE_CENTER
            R.id.action_3 -> PAGE_RIGHT
            R.id.action_4 -> PAGE_R18
            else -> -1
        }
    }

    private fun forceRefreshCurrentPage(itemId: Int) {
        val pageIndex = getPageIndex(itemId)
        if (pageIndex == -1) {
            return
        }
        forceRefreshPage(pageIndex)
    }

    private fun forceRefreshPage(pageIndex: Int) {
        when (val fragment = getBaseFragment(pageIndex)) {
            is FragmentLeft -> fragment.forceRefresh()
            is FragmentCenter -> fragment.forceRefresh()
            is FragmentRight -> fragment.forceRefresh()
            is FragmentViewPager -> fragment.forceRefresh()
        }
    }

    private fun getBaseFragment(pageIndex: Int): Fragment? {
        if (pageIndex < 0 || pageIndex >= baseFragments.size) {
            return null
        }
        return baseFragments[pageIndex]
    }

    private fun syncBottomNavigation(position: Int) {
        val itemId = getBottomNavigationItemId(position)
        if (itemId != View.NO_ID) {
            baseBind.navigationView.selectedItemId = itemId
        }
    }

    private fun getBottomNavigationItemId(pageIndex: Int): Int {
        return when (pageIndex) {
            PAGE_LEFT -> R.id.action_1
            PAGE_CENTER -> R.id.action_2
            PAGE_RIGHT -> R.id.action_3
            PAGE_R18 -> R.id.action_4
            else -> View.NO_ID
        }
    }

    private fun resolveDrawerDestination(itemId: Int): Intent? {
        return when (itemId) {
            R.id.nav_gallery -> TemplateActivity.newDownloadManagerIntent(this)
            R.id.nav_slideshow -> TemplateActivity.newHistoryIntent(this)
            R.id.nav_manage -> TemplateActivity.newSettingsIntent(this)
            R.id.nav_share -> TemplateActivity.newAboutIntent(this)
            R.id.main_page -> UActivity.newCurrentUserIntent(this)
            R.id.muted_list -> TemplateActivity.newMutedTagsIntent(this)
            R.id.nav_feature -> TemplateActivity.newFeatureIntent(this)
            R.id.nav_fans -> TemplateActivity.newFansIntent(this, Shaft.sUserModel.getUser().getId())
            R.id.illust_star -> TemplateActivity.newMyIllustBookmarksIntent(this)
            R.id.novel_star -> TemplateActivity.newMyNovelBookmarksIntent(this)
            R.id.watchlist -> TemplateActivity.newWatchlistIntent(this)
            R.id.novel_markers -> TemplateActivity.newNovelMarkersIntent(this)
            R.id.follow_user -> TemplateActivity.newMyFollowingIntent(this)
            else -> null
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.clear()
    }

    private fun initDrawerHeader() {
        val user = Shaft.sUserModel.getUser() ?: return
        Glide.with(mContext)
            .load(GlideUtil.getHead(user))
            .into(userHead!!)
        username?.text = user.getName()
        userEmail?.text = if (TextUtils.isEmpty(user.getMail_address())) {
            mContext.getString(R.string.no_mail_address)
        } else {
            user.getMail_address()
        }
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        if (handleOpenDrawerOnBack()) {
            return true
        }
        if (shouldInterceptBack(keyCode, event)) {
            exit()
            return true
        }
        return false
    }

    private fun handleOpenDrawerOnBack(): Boolean {
        if (!baseBind.drawerLayout.isDrawerOpen(GravityCompat.START)) {
            return false
        }
        baseBind.drawerLayout.closeDrawer(GravityCompat.START)
        return true
    }

    private fun shouldInterceptBack(keyCode: Int, event: KeyEvent): Boolean {
        return keyCode == KeyEvent.KEYCODE_BACK && event.repeatCount == 0
    }

    fun exit() {
        if (shouldFinishOnBackPress()) {
            finish()
            return
        }
        if (hasPendingDownloadTasks()) {
            showDownloadTaskExitDialog()
            return
        }
        recordPendingExit()
    }

    private fun shouldFinishOnBackPress(): Boolean {
        return System.currentTimeMillis() - mExitTime <= 2000
    }

    private fun hasPendingDownloadTasks(): Boolean {
        return Manager.get().content.size != 0
    }

    private fun showDownloadTaskExitDialog() {
        val builder = AlertDialog.Builder(mContext)
        builder.setTitle(getString(R.string.shaft_hint))
        builder.setMessage(mContext.getString(R.string.you_have_download_plan))
        builder.setPositiveButton(mContext.getString(R.string.sure)) { _: DialogInterface, _: Int ->
            stopDownloadsAndExit()
        }
        builder.setNegativeButton(mContext.getString(R.string.cancel), null)
        builder.setNeutralButton(getString(R.string.see_download_task)) { _, _ ->
            openDownloadManager()
        }
        builder.create().show()
    }

    private fun stopDownloadsAndExit() {
        Manager.get().stopAll()
        finish()
    }

    private fun openDownloadManager() {
        TemplateActivity.startDownloadManager(this)
    }

    private fun recordPendingExit() {
        Common.showToast(getString(R.string.double_click_finish))
        mExitTime = System.currentTimeMillis()
    }

    override fun onResume() {
        super.onResume()
        if (shouldRefreshDrawerHeader()) {
            initDrawerHeader()
            intent.removeExtra(EXTRA_REFRESH_DRAWER_HEADER)
        }
    }

    private fun shouldRefreshDrawerHeader(): Boolean {
        return intent.getBooleanExtra(EXTRA_REFRESH_DRAWER_HEADER, false)
    }

    override fun finish() {
        val currentPosition = baseBind.viewPager.currentItem
        Shaft.getMMKV().putInt(Params.MAIN_ACTIVITY_NAVIGATION_POSITION, currentPosition)
        super.finish()
    }

    private fun getNavigationInitPosition(): Int {
        val defaultPosition = 0
        val settingValue = Shaft.sSettings.getNavigationInitPosition()
        if (settingValue == NavigationLocationHelper.LATEST) {
            val latestPosition = Shaft.getMMKV().getInt(Params.MAIN_ACTIVITY_NAVIGATION_POSITION, 0)
            return if (latestPosition < baseFragments.size) latestPosition else defaultPosition
        }
        val navigationValue = NavigationLocationHelper.NAVIGATION_MAP.getOrDefault(settingValue, null)
            ?: return defaultPosition
        val clazz = navigationValue.getInstanceClass()
        for (i in baseFragments.indices) {
            val fragment = baseFragments[i]
            if (clazz == fragment.javaClass) {
                return i
            }
        }
        return defaultPosition
    }

    companion object {
        private const val REQUEST_WRITE_STORAGE = 1001
        private const val EXTRA_REFRESH_DRAWER_HEADER = "refreshDrawerHeader"
        private const val PAGE_LEFT = 0
        private const val PAGE_CENTER = 1
        private const val PAGE_RIGHT = 2
        private const val PAGE_R18 = 3

        @JvmStatic
        fun newIntent(context: Context, refreshDrawerHeader: Boolean): Intent {
            val intent = Intent(context, MainActivity::class.java)
            intent.putExtra(EXTRA_REFRESH_DRAWER_HEADER, refreshDrawerHeader)
            return intent
        }
    }
}
